File: examples/c/msreader.c
//
//  Reading from multiple sockets in C
//
#include <zmq.h>
#include <time.h>

int main (int argc, char *argv[])
{
    void *context;          //  ØMQ context for our process
    void *s1, *s2;

    //  Prepare our context and sockets
    context = zmq_init (1);

    //  Connect to task ventilator
    s1 = zmq_socket (context, ZMQ_PULL);
    zmq_connect (s1, "tcp://localhost:5557");

    //  Connect to weather server
    s2 = zmq_socket (context, ZMQ_SUB);
    zmq_connect (s2, "tcp://localhost:5556");
    zmq_setsockopt (s2, ZMQ_SUBSCRIBE, "10001 ", 6);

    //  Process messages from both sockets
    //  We prioritize traffic from the task ventilator
    while (1) {
        int rc;
        struct timespec t;
        zmq_msg_t message;

        //  Process any waiting tasks
        for (rc = 0; !rc; ) {
            zmq_msg_init (&message);
            if ((rc = zmq_recv (s1, &message, ZMQ_NOBLOCK)) == 0) {
                //  process task
            }
            zmq_msg_close (&message);
        }
        //  Process any waiting weather updates
        for (rc = 0; !rc; ) {
            zmq_msg_init (&message);
            if ((rc = zmq_recv (s2, &message, ZMQ_NOBLOCK)) == 0) {
                //  process weather update
            }
            zmq_msg_close (&message);
        }
        //  No activity, so sleep for 1 msec
        t.tv_sec = 0;
        t.tv_nsec = 1000000;
        nanosleep (&t, NULL);
    }
    return 0;
}

File: examples/PHP/msreader.php
<?php
/*
 *  Reading from multiple sockets
 *  This version uses a simple recv loop
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */

//  Prepare our context and sockets
$context = new ZMQContext();

//  Connect to task ventilator
$receiver = new ZMQSocket($context, ZMQ::SOCKET_PULL);
$receiver->connect("tcp://localhost:5557");

//  Connect to weather server
$subscriber = new ZMQSocket($context, ZMQ::SOCKET_SUB);
$subscriber->connect("tcp://localhost:5556");
$subscriber->setSockOpt(ZMQ::SOCKOPT_SUBSCRIBE, "10001");

//  Process messages from both sockets
//  We prioritize traffic from the task ventilator
while(true) {
	//  Process any waiting tasks
	try {
		for($rc = 0; !$rc;) {
			if($rc = $receiver->recv(ZMQ::MODE_NOBLOCK)) {
				// process task
			}
		}
	} catch (ZMQSocketException $e) {
		// do nothing 
	}
	
	try {
		//  Process any waiting weather updates
		for($rc = 0; !$rc;) {
			if($rc = $subscriber->recv(ZMQ::MODE_NOBLOCK)) {
				// process weather update
			}
		}
	} catch (ZMQSocketException $e) {
		// do nothing 
	}
		
	//  No activity, so sleep for 1 msec
	usleep(1);
}